### PR TITLE
initial commit implementing Hash unification attempt

### PR DESCRIPTION
**Pull request by dbmcclain**
incorporate Hex module and "proper" error handling...

fix debug config file for curve1174 and pbc example code

remove redundant "use hex"

fix up to return a Result

get "".hash() to work - must implement hasher and pass along. No useful results until hasher.result() is called.

Implement full hashing for publicly exposed types: G1, G2, PublicKey, SecretKey, etc...

Implement new Signature type to tag uses of G1

Implement new RVal type to tag uses of G2 in SAKKAI encrypted packets

### CODE REVIEW

**Comment by dbmcclain** [comment]
This PR incorporates changes needed by Roman for his block header definition.

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,293 @@
+{
```
Comment (rtsisyk) [inline .vscode/launch.json]: This file is unrelated, please exclude.

File: crypto/src/curve1174/fr.rs [hunk 2]
```diff
@@ -130,6 +130,21 @@ impl fmt::Display for Fr {
...
+
+impl Hashable for Fr {
+    fn hash(&self, state: &mut Hasher) {
+        match (*self).unscaled() {
+            Fr::Unscaled(U256(v)) => {
+                let lv = Lev32 { v64: v };
+                "Fr".hash(state);
+                unsafe { lv.v8 }.hash(state);
+            },
+            _ => cant_happen()
```
Comment (rtsisyk) [inline crypto/src/curve1174/fr.rs]: please don't forget change to unreachable!() as discussed.

File: crypto/src/hash/mod.rs [hunk 3]
```diff
@@ -46,8 +46,26 @@ impl Hash {
         self.0
     }
 
+    fn standard_hasher() -> Sha3_256 {
+        Sha3_256::new()
```
Comment (rtsisyk) [inline crypto/src/hash/mod.rs]: Probably it should return our Hasher instead of Sha3_256::new() in order to encapsulate Sha3_256 in one place and simplify from_vector() and from_str(0 below.

File: crypto/src/pbc/mod.rs [hunk 4]
```diff
@@ -37,6 +37,8 @@ use rust_libpbc;
 pub mod fast;
 pub mod secure;
 
+use hex;
+use std::prelude::v1::*;
```
Comment (rtsisyk) [inline crypto/src/pbc/mod.rs]: It shouldn't be required.

File: examples/curve1174.rs [hunk 5]
```diff
@@ -36,9 +36,16 @@ necessary, these are forcibly viewed through an "unsafe" transformation to/from
 ---------------------------------------------------------------------------- */
 extern crate stegos_crypto;
 
+use std::prelude::v1::*;
```
Comment (rtsisyk) [inline examples/curve1174.rs]: It shouldn't be required.

File: examples/curve1174.rs [hunk 6]
```diff
@@ -36,9 +36,16 @@ necessary, these are forcibly viewed through an "unsafe" transformation to/from
...
 use stegos_crypto::curve1174::*;
 
+extern crate lazy_static;
```
Comment (rtsisyk) [inline examples/curve1174.rs]: Unused

File: examples/pbc.rs [hunk 7]
```diff
@@ -33,9 +33,15 @@ use stegos_crypto::utils::*;
 
 use std::sync::Mutex;
 
+extern crate lazy_static;
```
Comment (rtsisyk) [inline examples/pbc.rs]: Unused.

File: crypto/src/curve1174/mod.rs [hunk 8]
```diff
@@ -31,7 +31,9 @@ use rand::prelude::*;
 use std::fmt;
 use std::mem;
 
+use hex;
 use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+use std::prelude::v1::*;
```
Comment (rtsisyk) [inline crypto/src/curve1174/mod.rs]: It shouldn't be required.

File: .vscode/launch.json (hunk 1 above)
Comment (dbmcclain) [reply .vscode/launch.json]: done

File: crypto/src/hash/mod.rs (hunk 3 above)
Comment (dbmcclain) [reply crypto/src/hash/mod.rs]: done

File: examples/pbc.rs (hunk 7 above)
Comment (dbmcclain) [reply examples/pbc.rs]: unused for now... will use shortly

File: examples/curve1174.rs (hunk 6 above)
Comment (dbmcclain) [reply examples/curve1174.rs]: unused for now... will use shortly

File: examples/curve1174.rs (hunk 5 above)
Comment (dbmcclain) [reply examples/curve1174.rs]: done
